k-distribution/k-tutorial/1_basic/02_basics/README.md

```diff
@@ -1 +1,212 @@
+types. Tokens, brackets, lists, macros, aliases, and anywhere productions will
+be covered in a later lesson, but this lesson does introduce us to constructors
+and functions. `Yellow()`, `Blue()`, `Banana()`, and `Blueberry()` are
+constructors. You can think of a constructor like a constructor for an
+algebraic data type, if you're familiar with a functional language. The data
+type itself is the sort that appears on the left of the `::=` operator. Sorts
+in K consist of uppercase identifiers.
+
+Constructors can have arguments, but these constructors do not have any
+arguments. We will cover the syntax of productions in detail in the next
+lesson, but for now, you can write a production with no arguments as an
+uppercase or lowercase identifier followed by the `()` operator. You can write
+a production with arguments by putting the sorts of the arguments in a
+comma-separated list inside the parentheses.
+
+A function is distinguished from a constructor by the presence of the
+**function** attribute. Attributes appear in a comma separated list between
+square brackets after any sentence, including both productions and rules.
+Various attributes with built-in meanings exist in K and will be discussed
+throughout the tutorial, but the user can also give any sentence a user-defined
+attribute using the same syntax by simply using an identifier for the attribute
+which is not reserved by K.
+
+## Rules, Matching, and Variables
+
+Functions in K are given definitions using rules. A rule begins with the `rule`
+keyword and contains at least one **rewrite operator**. The rewrite operator
+is represented by the syntax `=>`. The rewrite operator is one of the built-in
+productions in K, and we will discuss in more detail how it can be used in
+future lessons, but for now, you can think of a rule as consisting of a 
+**left-hand side** and a **right-hand side**, separated by the rewrite
+operator. On the left-hand side is the name of the function and zero or more
+**patterns**. On the right-hand side is another pattern. The meaning of the
+rule is relatively simple, having defined these components. If the function
+is called with arguments that **match** the patterns on the left-hand side,
+then the return value of the function is the pattern on the right-hand side.
+
+For example, in the above example, if the argument of the `colorOf` function
+is `Banana()`, then the return value of the function is `Yellow()`.
+
+So far we have introduced that a constructor is a type of pattern in K. We
+will introduce more complex patterns in later lessons, but there is one other
+type of basic pattern: the **variable**. A variable, syntactically, consists
+of an uppercase identifier. However, unlike a constructor, a variable will
+match **any** pattern with one exception: Two variables with the same name
+must match the same pattern.
+
+Here is a more complex example:
+
+```k
+module LESSON-02-D
+
+  syntax Container ::= Jar(Fruit)
```

Comment:
Make sure you mention, explicitly here, that `Jar(_)` is now a *constructor* which takes a single *argument*. Above I mention not to introduce multi-argument constructor explanation until they are introduced, so you should move that text I had you delete to down here instead. You should say "this is a single-argument constructor, but you can have multi-argument constructors by adding multiple sorts separated by commas".